Update metrics.md Changed /common/tls to /current/tls for metrics.crt, metrics.key & server.crt location

JohnHammell opened the pull request:
Changed: /var/snap/prometheus/common/tls/ to /var/snap/prometheus/current/tls/ as the location for the metrics.crt, metrics.key & server.crt files in Prometheus.

I was getting no data in Grafana after following this page.

Looked into my syslog and found the following message:

2024-05-11T03:08:48.449690+00:00 lxd-metrics prometheus.prometheus[2973]: ts=2024-05-11T03:08:48.448Z caller=manager.go:212 level=error component="scrape manager" msg="error creating new scrape pool" err="error creating HTTP client: unable to load specified CA cert /var/snap/prometheus/86/tls/server.crt: open /var/snap/prometheus/86/tls/server.crt: no such file or directory" scrape_pool=lxd

So I made a copy of server.crt in the /86/tls/ folder where the message states it was missing from and that error went away.

Then started seeing this next error regarding metrics.crt & metrics.key:

2024-05-11T03:10:31.336446+00:00 lxd-metrics prometheus.prometheus[3051]: ts=2024-05-11T03:10:31.335Z caller=manager.go:212 level=error component="scrape manager" msg="error creating new scrape pool" err="error creating HTTP client: unable to use specified client cert (/var/snap/prometheus/86/tls/metrics.crt) & key (/var/snap/prometheus/86/tls/metrics.key): open /var/snap/prometheus/86/tls/metrics.crt: no such file or directory" scrape_pool=lxd

Also moved metrics.crt & metrics.key into /86/tls/ and that second error went away.

As a test, I removed metrics.key, metrics.crt & server.crt from the original /common/tls/ location and that did not cause a problem.

It appears that Prometheus needs those .crt & .key files in the /var/snap/prometheus/86/tls folder rather than /var/snap/prometheus/common/tls , otherwise I get those errors about them being missing.

/current/ is a symlink to /86/ , so I figured there must be a reason for that and used /current/ in my edit instead of /86/

And, finally, in case this makes any difference, here is my setup:

Using Ubuntu 22.04 as my host LXD 5.21 server, I followed this page and setup Prometheus, Grafana, Loki & Promtail inside an Ubuntu 24.04 container. Proxied web access to Grafana through HAProxy.

JohnHammell commented:
Just to provide a little more context... On the same page where I'm suggesting this small change to the path for the certs, go to the following sectiont: https://documentation.ubuntu.com/lxd/en/stable-5.21/metrics/#configure-prometheus-to-scrape-from-lxd

There you'll see that the prometheus.yml config file is located at /var/snap/prometheus**/current/**prometheus.yml and in that file the certs are referenced with links relative to that same folder:

`      ca_file: 'tls/server.crt'
      cert_file: 'tls/metrics.crt'
      key_file: 'tls/metrics.key'`

Meaning it must be intended for them to be in:

/var/snap/prometheus/current/tls/server.crt
/var/snap/prometheus/current/tls/metrics.crt
/var/snap/prometheus/current/tls/metrics.key

Rather than in the /common/ folder.

Noticed these references on that page earlier today when I was working something else out and figured I should add that additional context into this PR.

tomponline commented:
@simondeziel please can you take a look at this, I'm concerned that `/var/snap/prometheus/current` is a temporary location and so making modifications to it (as the guide suggests) would be lost after a refresh to a new revision.

simondeziel commented:
@JohnHammell I share the same concerns as Tom, using the `current` directory is risky so I reported it: https://bugs.launchpad.net/prometheus-snap/+bug/2066910

To your point of making our documented procedure work, I think a symlink from `current/tls/ -> common/tls/` would be a better workaround. What do you think? If you agree, please add a link to the Launchpad bug right next to the symlink creation.

Thanks!

JohnHammell commented:
@simondeziel The actual location where Prometheus is looking for the .crt & .key files is /var/snap/prometheus/86/tls/  That's what was shown in my syslog error messages.

The `current` directory is already a symlink to `86`, so `current/tls` already goes to `86/tls`

Did you mean to symlink `common/tls` -> `current/tls`?

I'm not sure of the significance of the `86` directory, but if that is permanent we could alternatively symlink `common/tls` -> `86/tls`

Either way would do it to make the doc instructions work for placing those .crt & .key files in `common/tls` so they get to where Prometheus looks for them in `86/tls`

Another option I suppose would be to update the doc instructions to use the `86` directory instead of the `current` directory/symlink that I suggested.

I don't know which way is the preferred method but let me know and I'll make those changes.

simondeziel commented:
@JohnHammell I meant to have `/var/snap/prometheus/current/tls/` be a symlink to `/var/snap/prometheus/common/tls/`. This way, when `current` is refreshed, only the symlink needs to be recreated.

As for `current` already being a symlink, that's by design where it points to the currently active revision, `86` in your case but that's a moving target. Refreshing or reverting the snap, or switching channel would have `current` point somewhere else. That's why `common` feels like a better home for configs and keys.

ru-fu commented:
So if I understand correctly, the current documentation has the instructions how we would want them to work - we put the files into a persistent directory (`common`), and if the snap is updated, they are still used.
But the problem is that the snap doesn't find the files in this directory, because it uses a relative link to the current snap version (`86` or `current`).

If that understanding is correct, I agree with @simondeziel - adding a symlink as a workaround is the cleanest solution.
For the docs it means we can keep the instructions as they are and only add a fourth step to create a `tls` symlink in the `current` (resp. `86`) folder that redirects to `../common/tls`.
(Plus add a short explanation on why and add a link to the Launchpad bug.)

JohnHammell commented:
@simondeziel & @ru-fu  I probably don't have the best knowledge about how snaps are structured in general and also in regards to permanent, temporary & moving target type directories with snaps. But with your help I am better understanding these things now. :)

I now see how the symlink accomplishes what we need to do. The symlink being a sort of nested symlink is probably what was tripping me up a bit. I understand now that we want references to folders such as `86` to go through `current` since that will always be pointing to the current "moving target" folder.

I updated the PR changes with the following:
- Reverted my initial suggestion of using current/ in the doc back to common/
- And added in the suggested symbolic link so we can use common/tls as the location for the .crt & .key files. Launchpad bug link is directly above the symlink command as a comment.

Not sure if I should go further to explain more about the symlink or if that was enough. Let me know if more details about it should be added in to the doc.

ru-fu reviewed (approved):
Looks good to me, thanks!

tomponline commented:
@ru-fu is the cla check error valid?

ru-fu commented:
> @ru-fu is the cla check error valid?

Looks like the check is broken again ... 
It was green before: https://github.com/canonical/lxd/actions/runs/9122869012/job/25111815006

simondeziel reviewed (commented):
@JohnHammell the end result is good. Would you mind just squashing your 2 commits together rather than having the second undoing the changes of the first?

`git rebase -i main` should let you `squash` the 2 and then edit the resulting commit message.

Thanks!

JohnHammell commented:
> @JohnHammell the end result is good. Would you mind just squashing your 2 commits together rather than having the second undoing the changes of the first?
> 
> `git rebase -i main` should let you `squash` the 2 and then edit the resulting commit message.
> 
> Thanks!

Sure, no problem.

Tried doing the squash via the GitHub web UI, but looked all over and could not find the option to do that anywhere.

I'm not yet very fluent in git command line, but with the help of freecodecamp and another blog post I found, I believe I may have done the squash that you requested. The number of commits here is now 1, and the code view looks the way it should, so I think we might be good to go with it.

simondeziel reviewed (approved):
Awesome, LGMT, thanks again!
